python/src/functions/generate_treasury_report.py

```diff
@@ -0,0 +1,469 @@
+        files contain projects that we want to remove from the output file.
+        Then remove those files from the output file while simulateneously
+        updating the metadata (when you delete a row in the XLSX, the metadata
+        needs to be updated).
+    4) Open files in the uploadsByExpenditureCategory parameter. These files
+        contain projects that we want to add to the output file. There is some
+        complicated logic to handle cases where there are duplicate project
+        entries from the SAME agency. When that happens, we want to use the
+        latest entry, denoted by the UploadObject's createdAt data. OR if there
+        is already an entry in the output file, overwrite it since presumably
+        the new worksheet has the updated data.
+    5) Save the XLSX sheet, convert it to CSV and upload it, save metadata to
+        s3.
+
+    Args:
+        event: S3 Lambda event of type `s3:ObjectCreated:*`
+        context: Lambda context
+    """
+    structlog.contextvars.bind_contextvars(lambda_event={"step_function": event})
+    logger = get_logger()
+    logger.info("received new invocation event from step function")
+
+    s3_client: S3Client = boto3.client("s3")
+
+    project_code = os.getenv("PROJECT_USE_CODE")
+    if project_code == "1A":
+        project_use_code = ProjectType._1A
+    elif project_code == "1B":
+        project_use_code = ProjectType._1B
+    elif project_code == "1C":
+        project_use_code = ProjectType._1C
+    ProjectRowSchema = getSchemaByProject(VERSION, project_use_code)
+
+    organization: Dict[str, Any] = event["Records"][0]["organization"]
+    organization_id = organization.get("id")
+    current_reporting_period_id = organization.get("preferences").get(
+        "current_reporting_period_id"
+    )
+    user: Dict[str, Any] = event["Records"][0]["user"]
+
+    uploadsByExpenditureCategory = event.get("uploadsByExpenditureCategory", {})
+    file_info_list = uploadsByExpenditureCategory.get(project_use_code, {})
+    outdatedUploadsByExpenditureCategory = event.get(
+        "outdatedUploadsByExpenditureCategory", {}
+    )
+    outdated_file_info_list = outdatedUploadsByExpenditureCategory.get(
+        project_use_code, {}
+    )
+
+    # If the treasury report file exists, download it and store the data
+    # If it doesn't exist, download the output template
+
+    ### 1) Load the metadata
+    project_agency_id_to_row_map = get_existing_output_metadata(
+        s3_client=s3_client,
+        organization=organization,
+        user=user,
+        project_use_code=project_use_code,
+    )
+
+    ### 2) Download the existing output XLSX file. If it doesn't exist, download
+    #    the template file.
+    output_file = tempfile.NamedTemporaryFile()
+    if project_agency_id_to_row_map:
+        download_file(
+            s3_client,
+            f"treasuryreports/{organization_id}/{current_reporting_period_id}/{user.id}",
```

Comment:
@vshia one-small change here. Can we remove the `/{user.id}` prefix? The reports should be saved on the reporting period level and not be user-specific.

Apologies for not communicating this design change earlier. The change was made to ensure that the treasury report is the same regardless of who the user is for the entire organization.